### ISSUE TITLE
ResourceDef matching prefix "/" to "/foo/bar.baz" different in v4 beta

### ISSUE THREAD

**Issue by finnbear** [actix/actix-web#2389]
```rust
use actix_web::dev::ResourceDef;

fn main() {
    println!("{}", ResourceDef::root_prefix("").is_match("/build/bundle.js"));
    println!("{}", ResourceDef::root_prefix("/").is_match("/build/bundle.js"));
    println!("{}", ResourceDef::prefix("/").is_match("/build/bundle.js"));
    println!("{}", ResourceDef::prefix("/build").is_match("/build/bundle.js"));
    println!("{}", ResourceDef::prefix("/build/").is_match("/build/bundle.js"));
}

```

## Expected Behavior / `actix_web` 3.0.0
```
true
true
true
true
true
```

## Current Behavior / `actix_web` 4.0.0-beta.9
```
true
false // what is preventing my app from working
false
true
false
```

## Possible Solution

TBD (fix path parsing or justify why behavior is different and what mitigations can be taken)

If I had to guess, the cause is probably somewhere in: https://github.com/actix/actix-web/blob/a3806cde190ec27703576a91b470a6c8706cd5b1/actix-router/src/resource.rs#L1005-L1125

## Steps to Reproduce (for bugs)

Run the provided code with the two different versions of `actix_web`.

## Context
I'm trying to compile static files into my binary. [This](https://github.com/john01dav/actix-plus) worked fine with in `actix_web` 3.0.0. I [ported the package](https://github.com/finnbear/actix-plus) to support 4.0.0-beta.#, but it stopped matching paths. I narrowed the reason down to this issue.

## Your Environment

* Rust Version (I.e, output of `rustc -V`): `rustc 1.56.0-nightly (ae90dcf02 2021-08-09)`
* Actix Web Version: 4.0.0-beta.9

**Comment by finnbear** [comment]
Upon further investigation, it seems like the new path pattern matching is more sophisticated, and requires `""` or `"/*"` instead of `"/"` if the user wishes to match all files. This allowed me to write a mitigation for the change. It would be nice if this was documented better, though.

**Comment by aliemjay** [comment]
The docs are [already there](https://docs.rs/actix-web/4.0.0-beta.9/actix_web/dev/struct.ResourceDef.html#pattern-format-and-matching-behavior) :) but I think we should have more clear in the migration guide.

**Comment by finnbear** [comment]
I like the Migration Guide idea, but also consider putting an example of `*` in the docs. Or emphasize the possibility of using an empty string to match all paths (right now, the docs don't say exactly what an empty string means). I did read those docs (but would have had to click the RFC link to even get a hint about using an `*`).

**Comment by finnbear** [comment]
Adding to the list of questionable behavior that should be fixed or documented, I present the following expressions that both return true:
```rust
 ResourceDef::prefix("/core/*").is_match("/")
 ResourceDef::prefix("/core/*").is_match("/a")
```
This demonstrates how `*` at the end of a path can be very unintuitive.

**Comment by aliemjay** [comment]
tldr; just use `profix("/core")`

Hmm, while this is certainly a bug, please note that using `*` in this context is deprecated and deliberately not documented ([see deprecation notice](https://github.com/actix/actix-web/blob/a3806cde190ec27703576a91b470a6c8706cd5b1/actix-router/src/resource.rs#L1075)). `*` should now be used only in [named tail segments](https://docs.rs/actix-web/4.0.0-beta.9/actix_web/dev/struct.ResourceDef.html#tail-segments).

For your use case, you most probably want to use `prefix("/core")`; While `"/core"` and `"/core/*"` should match the same paths as a prefix, they are different in that the latter consumes the whole path when matching, not just the `/core` segment.

**Comment by aliemjay** [comment]
> Or emphasize the possibility of using an empty string to match all paths (right now, the docs don't say exactly what an empty string means).

Good point here! please feel free to open a PR for documentation , I may not be be the best to write docs :)

This should also be mentioned in `Scope` docs.

> I did read those docs (but would have had to click the RFC link to even get a hint about using an `*`).

You have probably confused this with the `*` in the ABNF notation in the RFC

**Comment by aliemjay** [comment]
@finnbear  submitted a PR to fix it in [your fork](https://github.com/finnbear/actix-plus/pull/1)
